clients/src/main/java/org/apache/kafka/common/utils/Time.java

```diff
@@ -86,4 +89,30 @@ default Timer timer(Duration timeout) {
         return timer(timeout.toMillis());
     }
 
+    /**
+     * Wait for a future to complete, or time out.
+     *
+     * @param future        The future to wait for.
+     * @param deadlineNs    The time in the future, in monotonic nanoseconds, to time out.
+     * @return              The result of the future.
+     * @param <T>           The type of the future.
+     */
+    default <T> T waitForFuture(
+        CompletableFuture<T> future,
+        long deadlineNs
+    ) throws TimeoutException, InterruptedException, ExecutionException  {
+        TimeoutException timeoutException = null;
+        while (true) {
+            long nowNs = nanoseconds();
+            if (deadlineNs <= nowNs) {
+                throw (timeoutException == null) ? new TimeoutException() : timeoutException;
+            }
+            long deltaNs = deadlineNs - nowNs;
+            try {
+                return future.get(deltaNs, TimeUnit.NANOSECONDS);
```

Comment:
it will not work if we are already past the deadline

Reply:
Thanks for following-up. However I double-checked the implementation of `CompletableFuture` and it throws a `TimeoutException` if `deltaNs` is <= 0. Am I missing something?

Reply:
The JavaDoc for CompletableFuture doesn't say what it does in case of negative timeouts. Therefore, it's a bad idea to rely on this behavior. Also, the subtraction you propose may overflow.

In any case this code was part of a different PR, #13153. I only included it here because I needed something from FutureUtils. Sorry for the confusion.

Reply:
Got it. Thanks.